x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/amazonbedrock/AmazonBedrockInferenceClient.java

```diff
@@ -0,0 +1,157 @@
+import com.amazonaws.services.bedrockruntime.model.AmazonBedrockRuntimeException;
+import com.amazonaws.services.bedrockruntime.model.ConverseRequest;
+import com.amazonaws.services.bedrockruntime.model.ConverseResult;
+import com.amazonaws.services.bedrockruntime.model.InvokeModelRequest;
+import com.amazonaws.services.bedrockruntime.model.InvokeModelResult;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.SpecialPermission;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.Strings;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.xpack.core.common.socket.SocketAccess;
+import org.elasticsearch.xpack.inference.services.amazonbedrock.AmazonBedrockModel;
+import org.joda.time.Instant;
+
+import java.security.AccessController;
+import java.security.PrivilegedExceptionAction;
+import java.util.Objects;
+
+/**
+ * Not marking this as "final" so we can subclass it for mocking
+ */
+public class AmazonBedrockInferenceClient extends AmazonBedrockBaseClient {
+
+    private static final int CLIENT_CACHE_EXPIRY_MINUTES = 5;
+    private static final long CACHE_EXPIRY_ADD_MS = (1000 * 60 * CLIENT_CACHE_EXPIRY_MINUTES);
+    private static final int DEFAULT_CLIENT_TIMEOUT_MS = 10000;
+
+    private final AmazonBedrockRuntime internalClient;
+    private volatile Instant expiryTimestamp;
+
+    public static AmazonBedrockBaseClient create(AmazonBedrockModel model, @Nullable TimeValue timeout) {
+        try {
+            return new AmazonBedrockInferenceClient(model, timeout);
+        } catch (Exception e) {
+            throw new ElasticsearchException("Failed to create Amazon Bedrock Client", e);
+        }
+    }
+
+    protected AmazonBedrockInferenceClient(AmazonBedrockModel model, @Nullable TimeValue timeout) {
+        super(model, timeout);
+        this.internalClient = createAmazonBedrockClient(model, timeout);
+        setExpiryTimestamp();
+    }
+
+    @Override
+    public ConverseResult converse(ConverseRequest converseRequest) throws ElasticsearchException {
+        try {
+            return internalClient.converse(converseRequest);
+        } catch (AmazonBedrockRuntimeException amazonBedrockRuntimeException) {
+            throw new ElasticsearchException(
+                Strings.format("AmazonBedrock converse failure: [%s]", amazonBedrockRuntimeException.getMessage()),
+                amazonBedrockRuntimeException
+            );
+        } catch (ElasticsearchException elasticsearchException) {
+            // just throw the exception if we have one
+            throw elasticsearchException;
+        } catch (Exception e) {
+            throw new ElasticsearchException("Amazon Bedrock client converse call failed", e);
+        }
+    }
+
+    @Override
+    public InvokeModelResult invokeModel(InvokeModelRequest invokeModelRequest) throws ElasticsearchException {
+        try {
+            return internalClient.invokeModel(invokeModelRequest);
```

Comment:
From what I can see in the docs this looks like a synchronous call: https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/java_bedrock-runtime_code_examples.html#amazon_titan_text_embeddings

In the `HttpClient` class we use the asynchronous variant of the client. We should probably use the async version of the amazon sdk here too.

https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/bedrockruntime/BedrockRuntimeAsyncClient.html

Hopefully that doesn't cause a lot of changes 🤞 maybe we could pass the listener in here and the `converse` method?

One thing to keep in mind when using the async client is which thread the response is returned on. When I was reading the apache async client docs they explicitly say to not do any heavy operations on the result thread so we intentionally jump to a utility thread for each response before calling `listener.onResponse()`.

Reply:
Unfortunately, the async version appears to be only in version 2 of the SDK - we'e using using version 1 (version `1.12.740` to be exact) to stay compatible with the AWS SDK v1 used in the respository-s3 pieces (version `1.12.270`)... :/. - otherwise, I'd 100% agree with you on this.

Reply:
Wooza - that was a trip ;). But - I think I got this in an async manner with proper listeners for the Amazon runtime responses... and, I think it's much more solid now with this... I'll have another commit shortly.

Reply:
@jonathan-buttner - I've added in the Async pieces and did some cleanups... there's still a number of your comments above I need to address.

Reply:
Thanks for doing that!

Do you know if we use SDK1 in x-pack core? If we don't, I think we might be able to update it separate from other plugins 🤔 . If it's in core I don't think we can have separate versions in core and in another plugin. Looks like SDK1 is EOL next year: https://aws.amazon.com/blogs/developer/announcing-end-of-support-for-aws-sdk-for-java-v1-x-on-december-31-2025/

How about we create an issue to remind myself to update the SDK to version 2 if we can.

Reply:
> Do you know if we use SDK1 in x-pack core?

I know it's used in the [respository-s3 module](https://github.com/elastic/elasticsearch/tree/main/modules/repository-s3), but aside from that, I can't find it used anywhere else (well, besides this PR) - and because this is in a module, I thought it best to not try and use v2 at the moment... 😞 

> How about we create an issue to remind myself to update the SDK to version 2 if we can.

Great idea - which repo / system? (Github / Jira? Which team?)

Reply:
> Great idea - which repo / system? (Github / Jira? Which team?)

The ML team uses the elasticsearch github repo mainly, how about that and we tag it with `:ml` and `:EnterpriseSearch/Application`

Reply:
Sounds good - I'll set it up